integration-cli/docker_cli_cp_from_container_test.go

```diff
@@ -144,6 +144,33 @@ func (s *DockerSuite) TestCpFromSymlinkDestination(c *check.C) {
 	// The file should have the contents of "file2" now.
 	c.Assert(fileContentEquals(c, cpPath(tmpDir, "file2"), "file2\n"), checker.IsNil)
 
+	/*
+	// Now, we need to test copying a file (file2) to a symlink directory with a new
+	// target file name.
+	dstPath = cpPath(tmpDir, "symlinkToDir4", "newname")
+	c.Assert(runDockerCp(c, srcPath, dstPath, nil), checker.IsNil)
+
+	// Verify that newfile has the contents of file2 in the symlink directory
+	c.Assert(fileContentEquals(c, cpPath(tmpDir, "symlinkToDir4", "newname"), "file2\n"), checker.IsNil)
+
+	// Verify that newfile has the contents of file2 in the real directory
+	c.Assert(fileContentEquals(c, cpPath(tmpDir, "dir4", "newname"), "file2\n"), checker.IsNil)
+
+	//Verify that copying to a symlinked symlink directory works.
+	srcPathFile3 := containerCpPath(containerID, "/file3")
+	dstPath = cpPath(tmpDir, "symlinkToDir4symlink", "newname2")
+	c.Assert(runDockerCp(c, srcPathFile3, dstPath, nil), checker.IsNil)
+
+	// Verify that newfile2 has the contents of file3 in symlinked symlink directory
+	c.Assert(fileContentEquals(c, cpPath(tmpDir, "symlinkToDir4symlink", "newname2"), "file3\n"), checker.IsNil)
+
+	// Verify that newfile2 has the contents of file3 in the symlinked directory
+	c.Assert(fileContentEquals(c, cpPath(tmpDir, "symlinkToDir4", "newname2"), "file3\n"), checker.IsNil)
+
+	// Verify that newfile2 has the contents of file3 in the real directory (dir4)
+	c.Assert(fileContentEquals(c, cpPath(tmpDir, "symlinkToDir4", "newname2"), "file3\n"), checker.IsNil)
+	*/
+
```

Comment:
Perhaps a unit test for `CopyInfoDestinationPath` would be enough? A quick draft;

```patch
From 5bec8cc8b4f1b89a1326d56d3040da54abf6e604 Mon Sep 17 00:00:00 2001
From: Sebastiaan van Stijn <github@gone.nl>
Date: Tue, 8 Aug 2017 17:20:53 +0200
Subject: [PATCH] Add unit tests

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>
---
 pkg/archive/changes_test.go   | 54 +++++++++++++++++++++++--------------------
 pkg/archive/copy_unix_test.go | 27 ++++++++++++++++++++++
 2 files changed, 56 insertions(+), 25 deletions(-)

diff --git a/pkg/archive/changes_test.go b/pkg/archive/changes_test.go
index c5d1629e7..25294652b 100644
--- a/pkg/archive/changes_test.go
+++ b/pkg/archive/changes_test.go
@@ -45,36 +45,40 @@ type FileData struct {
 	path        string
 	contents    string
 	permissions os.FileMode
+	expected    CopyInfo
 }
 
 func createSampleDir(t *testing.T, root string) {
 	files := []FileData{
-		{Regular, "file1", "file1\n", 0600},
-		{Regular, "file2", "file2\n", 0666},
-		{Regular, "file3", "file3\n", 0404},
-		{Regular, "file4", "file4\n", 0600},
-		{Regular, "file5", "file5\n", 0600},
-		{Regular, "file6", "file6\n", 0600},
-		{Regular, "file7", "file7\n", 0600},
-		{Dir, "dir1", "", 0740},
-		{Regular, "dir1/file1-1", "file1-1\n", 01444},
-		{Regular, "dir1/file1-2", "file1-2\n", 0666},
-		{Dir, "dir2", "", 0700},
-		{Regular, "dir2/file2-1", "file2-1\n", 0666},
-		{Regular, "dir2/file2-2", "file2-2\n", 0666},
-		{Dir, "dir3", "", 0700},
-		{Regular, "dir3/file3-1", "file3-1\n", 0666},
-		{Regular, "dir3/file3-2", "file3-2\n", 0666},
-		{Dir, "dir4", "", 0700},
-		{Regular, "dir4/file3-1", "file4-1\n", 0666},
-		{Regular, "dir4/file3-2", "file4-2\n", 0666},
-		{Symlink, "symlink1", "target1", 0666},
-		{Symlink, "symlink2", "target2", 0666},
-		{Symlink, "symlink3", root + "/file1", 0666},
-		{Symlink, "symlink4", root + "/symlink3", 0666},
-		{Symlink, "dirSymlink", root + "/dir1", 0740},
-	}
+		{filetype: Regular, path: "file1", contents: "file1\n", permissions: 0600},
+		{filetype: Regular, path: "file2", contents: "file2\n", permissions: 0666},
+		{filetype: Regular, path: "file3", contents: "file3\n", permissions: 0404},
+		{filetype: Regular, path: "file4", contents: "file4\n", permissions: 0600},
+		{filetype: Regular, path: "file5", contents: "file5\n", permissions: 0600},
+		{filetype: Regular, path: "file6", contents: "file6\n", permissions: 0600},
+		{filetype: Regular, path: "file7", contents: "file7\n", permissions: 0600},
+		{filetype: Dir, path: "dir1", contents: "", permissions: 0740},
+		{filetype: Regular, path: "dir1/file1-1", contents: "file1-1\n", permissions: 01444},
+		{filetype: Regular, path: "dir1/file1-2", contents: "file1-2\n", permissions: 0666},
+		{filetype: Dir, path: "dir2", contents: "", permissions: 0700},
+		{filetype: Regular, path: "dir2/file2-1", contents: "file2-1\n", permissions: 0666},
+		{filetype: Regular, path: "dir2/file2-2", contents: "file2-2\n", permissions: 0666},
+		{filetype: Dir, path: "dir3", contents: "", permissions: 0700},
+		{filetype: Regular, path: "dir3/file3-1", contents: "file3-1\n", permissions: 0666},
+		{filetype: Regular, path: "dir3/file3-2", contents: "file3-2\n", permissions: 0666},
+		{filetype: Dir, path: "dir4", contents: "", permissions: 0700},
+		{filetype: Regular, path: "dir4/file3-1", contents: "file4-1\n", permissions: 0666},
+		{filetype: Regular, path: "dir4/file3-2", contents: "file4-2\n", permissions: 0666},
+		{filetype: Symlink, path: "symlink1", contents: "target1", permissions: 0666},
+		{filetype: Symlink, path: "symlink2", contents: "target2", permissions: 0666},
+		{filetype: Symlink, path: "symlink3", contents: root + "/file1", permissions: 0666},
+		{filetype: Symlink, path: "symlink4", contents: root + "/symlink3", permissions: 0666},
+		{filetype: Symlink, path: "dirSymlink", contents: root + "/dir1", permissions: 0740},
+	}
+	provisionSampleDir(t, root, files)
+}
 
+func provisionSampleDir(t *testing.T, root string, files []FileData) {
 	now := time.Now()
 	for _, info := range files {
 		p := path.Join(root, info.path)
diff --git a/pkg/archive/copy_unix_test.go b/pkg/archive/copy_unix_test.go
index 4d5ae79cd..fc17ae010 100644
--- a/pkg/archive/copy_unix_test.go
+++ b/pkg/archive/copy_unix_test.go
@@ -15,6 +15,8 @@ import (
 	"path/filepath"
 	"strings"
 	"testing"
+
+	"github.com/stretchr/testify/assert"
 )
 
 func removeAllPaths(paths ...string) {
@@ -976,3 +978,28 @@ func TestCopyCaseJFSym(t *testing.T) {
 		t.Fatal(err)
 	}
 }
+
+func TestCopyInfoDestinationPathSymlink(t *testing.T) {
+	tmpDir, _ := getTestTempDirs(t)
+	defer removeAllPaths(tmpDir)
+
+	root := strings.TrimRight(tmpDir, "/") + "/"
+	files := []FileData{
+		{filetype: Regular, path: "file1", permissions: 0600, expected: CopyInfo{Path: root + "file1", Exists: true}},
+		{filetype: Dir, path: "dir1", permissions: 0700, expected: CopyInfo{Path: root + "dir1", Exists: true, IsDir: true}},
+		{filetype: Regular, path: "dir1/file1", permissions: 0600, expected: CopyInfo{Path: root + "dir1/file1", Exists: true}},
+		{filetype: Symlink, path: "symlink1", contents: "noSuchTarget", permissions: 0600, expected: CopyInfo{Path: root + "noSuchTarget", Exists: false}},
+		{filetype: Symlink, path: "symlink2", contents: "file1", permissions: 0600, expected: CopyInfo{Path: root + "file1", Exists: true}},
+		{filetype: Symlink, path: "symlink3", contents: root + "file1", permissions: 0600, expected: CopyInfo{Path: root + "file1", Exists: true}},
+		{filetype: Symlink, path: "symlink4", contents: root + "symlink3", permissions: 0600, expected: CopyInfo{Path: root + "file1", Exists: true}},
+		{filetype: Symlink, path: "dirSymlink", contents: root + "dir1", permissions: 0600, expected: CopyInfo{Path: root + "dir1", Exists: true, IsDir: true}},
+	}
+	provisionSampleDir(t, tmpDir, files)
+
+	for _, info := range files {
+		p := filepath.Join(tmpDir, info.path)
+		ci, err := CopyInfoDestinationPath(p)
+		assert.NoError(t, err)
+		assert.Equal(t, info.expected, ci)
+	}
+}
-- 
2.13.1
```